### PR TITLE
[client-go] Renegotiate Streaming Content-Type

### PR DESCRIPTION
**Pull request by DirectXMan12**
REST clients in client-go have two content type fields -- the "accepted"
content types, and the default content type.  When we set up
serializers and such, we set up default decoders, etc for the default
content type, and set up a "renegotiation function" to renegotiate in
case the server sends some other content type.

However, in the case of streaming, we don't do renegotiation properly --
we assume whatever the server sent is compatible with our default stream
framing/decoding.

This is fine an dandy when it actually is compatible, and is generally
fine an dandy when using the JSON content type as the default, because
most things support JSON.  However, if you set your default up to, say,
protobuf, and then try to use that REST config to construct a REST
client for talking to the custom resources API server, you're going to
have a bad time, since it doesn't support protobuf, only JSON.

This in turn causes weird behavior like watches starting succesfully,
but never actuall firing any watch events.

To fix this, this commit introduces a new renegotiation function for
streaming serialization and framing, and checks to see if it needs to be
used.

**Release note**:
```release-note
Fix content-type negotiation in streaming watch requests in client-go
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/rest/request.go [hunk 1]
```diff
@@ -532,9 +532,29 @@ func (r *Request) Watch() (watch.Interface, error) {
...
+
+	// check if we need to renegotiate for different content types (like done in
+	// transformResponse for non-streaming requests)
+	contentType := resp.Header.Get("Content-Type")
+	if len(contentType) > 0 && len(r.content.ContentType) > 0 && contentType != r.content.ContentType {
+		mediaType, params, err := mime.ParseMediaType(contentType)
+		if err != nil {
+			return nil, err
+		}
+		normalDecoder, streamingSerializer, framer, err = r.serializers.RenegotiatedStreaming(mediaType, params)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/rest/request.go]: this assumes RenegotiatedStreaming is non-nil?

File: staging/src/k8s.io/client-go/rest/client.go [hunk 2]
```diff
@@ -200,6 +201,14 @@ func createSerializers(config ContentConfig) (*Serializers, error) {
 	if info.StreamSerializer != nil {
 		s.StreamingSerializer = info.StreamSerializer.Serializer
 		s.Framer = info.StreamSerializer.Framer
+		s.RenegotiatedStreaming = func(contentType string, params map[string]string) (runtime.Decoder, runtime.Serializer, runtime.Framer, error) {
```
[the remark this answers is not in the document]
Comment (DirectXMan12) [reply staging/src/k8s.io/client-go/rest/client.go]: yep, will do.  Wanted to make sure there wasn't something I missed with the approach.

File: staging/src/k8s.io/client-go/rest/request.go (hunk 1 above)
Comment (DirectXMan12) [reply staging/src/k8s.io/client-go/rest/request.go]: the code seemed to assume StreamingSerializer was non-nil before.  I'll add a check in to be safe.

**Comment by wenjiaswe** [comment]
cc @yliaog

**Comment by liggitt** [comment]
sorry for letting this fall through the cracks

seems related to https://github.com/kubernetes/kubernetes/pull/67803#discussion_r225610529

**Comment by liggitt** [comment]
cc @jsafrane @smarterclayton

**Comment by DirectXMan12** [comment]
I'll try and rebase this soonish

**Comment by liggitt** [comment]
/close 

in favor of #84692

**Pull request closed without merging**